File: 3.recursion/509.FibonacciNumber/main.go
package main

// https://leetcode.com/problems/fibonacci-number/submissions/1626008400/

func fib(n int) int {
	if n < 2 {
		return n
	}

	return fib(n-1) + fib(n-2)
}

func fib2(n int) int {
	fibos := []int{0, 1}
	for i := 2; i <= n; i++ {
		fibos = append(fibos, fibos[i-1]+fibos[i-2])
	}

	return fibos[n]
}

func fib3(n int) int {
	if n < 2 {
		return n
	}
	prevNum := 1
	num := 1

	for n > 2 {
		temp := num
		num += prevNum
		prevNum = temp
		n--
	}

	return num
}

File: 3.recursion/509.FibonacciNumber/main.py
# https://leetcode.com/problems/fibonacci-number/

# TC: O(n), SC: O(n)
class Solution:
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        fibs = [0,1]
        for i in range(2, n+1):
            fibs.append(fibs[i-1] + fibs[i-2])
        return fibs[n]
    
# dynmaic programing
# TC: O(n): SC: O(n)
class Solution:
    def fib(self, n: int) -> int:
        dp = [-1] * (n + 1)
        def _fib(n):
            nonlocal dp
            if n <= 1:
                return n
            if dp[n] != -1:
                return dp[n]
            dp[n] = _fib(n-1) + _fib(n-2)
            print(dp)
            return dp[n]
        return _fib(n)
    